src/v/datalake/schema_parquet.cc

```diff
@@ -0,0 +1,225 @@
+        };
+    }
+};
+
+serde::parquet::schema_element
+struct_to_parquet(const iceberg::struct_type& schema);
+serde::parquet::schema_element
+field_to_parquet(const iceberg::nested_field& field);
+
+struct type_converting_visitor {
+    serde::parquet::schema_element
+    operator()(const iceberg::primitive_type& type) {
+        auto res = std::visit(primitive_type_converting_visitor{}, type);
+        res.repetition_type
+          = required ? serde::parquet::field_repetition_type::required
+                     : serde::parquet::field_repetition_type::optional;
+        return res;
+    }
+
+    serde::parquet::schema_element
+    operator()(const iceberg::struct_type& type) {
+        return struct_to_parquet(type);
+    }
+
+    serde::parquet::schema_element operator()(const iceberg::list_type& type) {
+        /**
+         * Lists in parquet are encoded as:
+         *
+         * <list-repetition> group <name> (LIST) {
+         *   repeated group list {
+         *     <element-repetition> <element-type> element;
+         *    }
+         * }
+         */
+        serde::parquet::schema_element res;
+        res.logical_type = serde::parquet::list_type{};
+        res.repetition_type
+          = required ? serde::parquet::field_repetition_type::required
+                     : serde::parquet::field_repetition_type::optional;
+
+        serde::parquet::schema_element list_wrapper;
+        list_wrapper.repetition_type
+          = serde::parquet::field_repetition_type::repeated;
+        list_wrapper.path.emplace_back("list");
+        auto element = field_to_parquet(*type.element_field);
+        element.path.emplace_back("element");
+        list_wrapper.children.push_back(std::move(element));
+        res.children.push_back(std::move(list_wrapper));
+        return res;
+    }
+
+    serde::parquet::schema_element operator()(const iceberg::map_type& type) {
+        /*
+         * Maps in parquet are encoded as:
+         * <map - repetition> group<name>(MAP) {
+         *     repeated group key_value {
+         *         required<key - type> key;
+         *         <value - repetition><value - type> value;
+         *     }
+         * }
+         */
+        serde::parquet::schema_element res;
+        serde::parquet::schema_element map_wrapper;
+        map_wrapper.repetition_type
+          = serde::parquet::field_repetition_type::repeated;
+        map_wrapper.path.emplace_back("key_value");
+        map_wrapper.children.reserve(2);
+        auto key_element = field_to_parquet(*type.key_field);
+        key_element.repetition_type
+          = serde::parquet::field_repetition_type::required;
+        key_element.path.emplace_back("key");
+        map_wrapper.children.push_back(std::move(key_element));
+        auto value_element = field_to_parquet(*type.value_field);
+        value_element.path.emplace_back("value");
+        value_element.repetition_type
+          = serde::parquet::field_repetition_type::optional;
```

Comment:
Values can be required or optional - can we make sure we correctly convert this.